cmd/client.go

```diff
@@ -0,0 +1,151 @@
+func (client *CloudproberClient) CloseConn() {
+	client.clientMux.Lock()
+	defer client.clientMux.Unlock()
+
+	err := client.conn.Close()
+	if err != nil {
+		glog.Errorf("Cloudprober gRPC client could not close connection, err: %v", err)
+	}
+}
+
+// addProbeFromConfig adds a probe to Cloudprober via the gRPC client.
+// Parameters:
+// - probePb: - This probe config must be unmarshalled before being passed as an argument.
+//			 - This probe type must be registered as an extension.
+// Returns:
+// TODO: Include details of these errors in the error message.
+// - error:
+//         - Code 3, InvalidArgument: probe config cannot be nil
+//         - error parsing regexp [...]: the machine name for this probe to run on does not compile as a regexp.
+//         - Code 6, AlreadyExists: probe [name] is already defined - the probe has already been added
+//         - Code 2, Unknown:
+//		- only one probe extension is allowed per probe, got %d extensions
+//			-> More than one probe extension type has been registered using cpprobes.RegisterProbeType()
+//		- no probe extension in probe config
+//			-> The probe proto does not have an extension registered.
+//		- no probes registered for the extension: [extension code]
+//			-> The proto has a probe extension, but there is no probe extension registered matching that extension code.
+//		- proto.GetExtension() error: extension missing, invalid, incomplete or proto cannot be extended.
+//			-> The proto.SetExtension() method must be called and supplied with valid data.
+//		- unregistered user defined probe: [probe name]
+//			-> The probe type has not been registered as a user defined probe.
+//		- unknown probe type: [probe type]
+//			-> The probe type does not match any expected types (including user-defined and extension).
+//		- probe.Init() error: an error occurred when the probe, created from the config passsed, was initialised.
+//			-> There was a problem in the Init() function of the probe type passed.
+//		- options.BuildProbeOptions() error: an error occurred when building the options for this probe from the config supplied.
+//			-> The options could not be built from the config supplied.
+func (client *CloudproberClient) addProbeFromConfig(ctx context.Context, probePb *configpb.ProbeDef) error {
+	// No mutex locking handled here as this is a private method called by
+	// a public method which handles locking.
+	_, err := client.client.AddProbe(ctx, &proberpb.AddProbeRequest{ProbeConfig: probePb})
+	return err
+}
+
+// RegisterAndAddProbe registers a probe type and adds the probe to Cloudprober to be run.
+// Parameters:
+// - probePb: - This probe config must be unmarshalled before being passed as an argument.
+// Returns:
+// - error:
+//	      - See addProbeFromConfig errors.
+func (client *CloudproberClient) RegisterAndAddProbe(extensionNumber int, ctx context.Context, probePb *configpb.ProbeDef, hermesProbeToAdd cpprobes.Probe) error {
```

Comment:
The context must be the first argument.

Reply:
I have made this change here and elsewhere where required.